### ISSUE TITLE
Passwordless Login

### ISSUE THREAD

**Issue by githubkusi** [sddm/sddm#751]
Please allow a passwordless login. This is a common usecase for family members. It has been already reported a few times (e.g #2, #577, #505), but the issues have always been closed.

I am not talking about autologin here, but a passwordless login for a second or third user

thanks!

**Comment by Azelphur** [comment]
I too am waiting on this feature, to give an example use case, you could have a computer which is shared between children, passwords on login are an unwanted hinderence, so the ideal answer would be when you click on one of the childrens accounts, it simply logs straight in, without requesting a password. You would of course want to be able to specify this on a per-account basis, so that you could secure an administration account.

**Comment by szg0000** [comment]
Autologin already implemented in sddm.
Follow this as root:
```
sddm --help
sddm --example-config
sddm --example-config > /etc/sddm.conf
```
Now you can modify /etc/sddm.conf. You need to modify only 2 lines for autologin.

```
# Name of session file for autologin session (if empty try last logged in)
Session=lxqt.desktop

# Username for autologin session
User=
```
`Session` must be "lxqt.desktop" `User` must be your username for autologin.

**Comment by mirh** [comment]
Is it SO hard to understand the difference between autologin and *NO PASSWORD*?
What Windows does: that's the aim.

**Comment by Azelphur** [comment]
@szg0000 auto login and passwordless login are different, what we want is...

1) Computer boots, sddm starts
2) The user is presented with the normal sddm display, and is able to select an account
3) When an account is selected, if that account is marked as passwordless, the desktop environment will be loaded, without the user entering a password.

To be clear, we know auto login already exists, it is entirely different to passwordless login and it does not allow us to select between users when sddm starts, and instead throws us straight into the desktop environment for the specified user. This is not what we want.

Hope this clears it up, let me know if you need further clarification :)

**Comment by Palando** [comment]
I think, it is so hard to understand the difference. Anyway I misunderstood it too. :-)

In linux users can have an empty password. (root can set it with "passwd -d user".) But then in SDDM you can not login. May be it would better to directly login without a password request in that case?

**Comment by Azelphur** [comment]
@Palando that sounds good to me

**Comment by ghmitch** [comment]
I just want to comment that I have used this KDM feature for years and I am wanting it included in SDDM as well.  Autologin does the job if you only have one user involved, but with multiple users you need the "passwordless login" feature whereby you can check off users allowed to login locally without entering a password.  With KDM the result is that they can just double click on their icon and THEN begin the autologin process.  But the SDDM screen is still needed as an initial selector screen to determine which user needs to autologin at any particular point in time.  That is why they are two different, although partially overlapping, features.  That said, I am really liking SDDM as a replacement for KDM.  KDM was robust but forever buggy, while SDDM is about the cleanest piece of software I have ever seen.  I have some other minor problems with SDDM but I am pretty certain they can all be solved with theming.  But not passwordless login.

**Comment by ktecho** [comment]
Is there any workaround to make this work, editing files by hand or whatever? Thanks!

**Comment by sleep-walker** [comment]
you can configure it on PAM level - see
https://wiki.archlinux.org/index.php/LightDM#Enabling_interactive_passwordless_login

**Comment by Azelphur** [comment]
@sleep-walker tried the PAM solution, it kinda works. SDDM still prompts for a password, but you can just press login, or enter any password and it'll work.

It's a step in the right direction, but I'd like for SDDM to recognize that a password isn't required and just log you in, rather than prompt for one.

**Comment by jcrben** [comment]
@szg0000 I guess for that Session value you take the desktop file in `/usr/share/xsessions` per https://github.com/sddm/sddm/blob/ef7e0812dab3d2e56b2ad388e6c45ca5c84391d7/src/common/Configuration.h#L65 - in my case with Plasma it was `Session=plasma.desktop`

**Comment by NatTuck** [comment]
Optimally, "passwordless login" shouldn't be tied to the account password at all. There should literally be a "passwordless local login" checkbox in the config for each user account, and that should have the effect of allowing a person at the SDDM login screen to select log in to that account without entering a password.

Consider the following case:

 * Alice is the administrator. She needs a password to log in.
 * Bob is an adult. He also wants his account password protected.
 * Carol, Dave, and Erin are kids. Anyone should be able to click on their account name at the login screen and log in.

 * Alice wants access to the machine by ssh and some random remote desktop app. The remote desktop app uses password auth and doesn't special-case blank passwords.
 * This machine is always on, and on an open WiFi network, so allowing any user to log in remotely with a blank password is a problem.

**Comment by xxlaefxx** [comment]
Hi guys! It's very-very insecure, but if you are really need it... 
I needed in passwordless login to make "kiosk" from my old nettop. List of dependencies for Lightdm looks like a moster, therefore I've choosen sddm.
After several hours of googling it hit me: I can write all logins/passwords into SDDM theme.
My button to start session with id "0" with user "laef" and "supersecret" password:
```
Button {
    id: loginButton0
    height: 100
    width: 100
        Image {
            id: kodiImage
            anchors.fill: parent
            source: "kodi.png"
        }
     text: " "
     color: "transparent"
     onClicked: sddm.login("laef", "supersecret", 0)
     KeyNavigation.backtab: loginButton1; KeyNavigation.tab: loginButton1
}
```
Yes, Images instead of text looks good and you can create the same buttons for another sessions/users.

Hope it will be helpful for someone

**Comment by sebastian-de** [comment]
I just found this issue, the solution is quite simple: Add users that should be allowed to login without a password to group `nopasswdlogin` (create it, if necessary).

uncomment the following line in `/etc/pam.d/sddm`
`auth    sufficient      pam_succeed_if.so user ingroup nopasswdlogin`

Works fine for me.

**Comment by guss77** [comment]
@sebastian-de it works well, thanks for the suggestion. The user experience, though, is not the greatest: the password prompt is displayed but any text typed into it - including the empty string (i.e. just hit ENTER) will log in.

Ideally, SDDM could detect that no password is needed and will just offer a login button without a password prompt - like LightDM used to do.

**Comment by sleep-walker** [comment]
this discussion is going in circles :b - see https://github.com/sddm/sddm/issues/751#issuecomment-346929324

**Comment by likewhoa** [comment]
> this discussion is going in circles :b - see [#751 (comment)](https://github.com/sddm/sddm/issues/751#issuecomment-346929324)

The problem here is that the password field is still visible and the request is to hide this when users are part of the nopasswdlogin group as to not confuse the end user. 

The use-case is simple, users in the 'nopasswdlogin' group shouldn't have to see a password text field since they would be allowed to login anyways. You should just be able to select a user and click login or hit the enter key.

**Comment by likewhoa** [comment]
> I too am waiting on this feature, to give an example use case, you could have a computer which is shared between children, passwords on login are an unwanted hinderence, so the ideal answer would be when you click on one of the childrens accounts, it simply logs straight in, without requesting a password. You would of course want to be able to specify this on a per-account basis, so that you could secure an administration account.



> Please allow a passwordless login. This is a common usecase for family members. It has been already reported a few times (e.g #2, #577, #505), but the issues have always been closed.
> 
> I am not talking about autologin here, but a passwordless login for a second or third user
> 
> thanks!

I am building a liveusb for Gentoo Linux at the moment and in the past I used the method mentioned above with PAM in order to allow the "gentoo" user to login with a blank password but this time around I wanted to just have the user field shown and so I modified the elarun theme in sddm to do just that. You can find the modified theme here  http://dpaste.com/1A2KRWJ

It basically does a login of the user "gentoo" without a password using.
``
sddm.login(user_entry.text, "", sessionIndex)
``
I removed all references to pw_entry and also the password field box. This allow the end user to just hit enter while on the user field box or by just clicking on the login icon. It's a better UX from my point of view.

**Comment by R0bur** [comment]
> uncomment the following line in `/etc/pam.d/sddm`
> `auth sufficient pam_succeed_if.so user ingroup nopasswdlogin`

This solution allows passwordless login for the members of "nopasswdlogin" group regardless of whether their passwords are blank or not. 

Another way is to replace "nullok_secure" with "nullok" in the file "/etc/pam.d/common_auth" as described here: https://unix.stackexchange.com/questions/316657/why-cant-su-be-used-to-switch-to-users-with-empty-password
It gives more predictable behaviour and doesn't requires "passwordless" group membership .

**Comment by tpecar** [comment]
Stumbled upon this same issue and checked the sources.

It seems that this was half-implemented a long time ago
https://github.com/sddm/sddm/blame/f9a0648ff3f841b68a07c139c3540b9e75bfd279/src/greeter/UserModel.cpp#L88

The default theme uses this to determine whether it should display the password box or the "press to login" prompt (confusingly, it doesn't respond to mouse clicks and still requires you to press the enter key)

https://github.com/sddm/sddm/blob/b304011c26cb16691ca8f3dcffe1712ac542e905/src/greeter/theme/Main.qml#L82
https://github.com/sddm/sddm/blob/6a683435fd5858fae8ae0d70db780d767b6886cc/components/2.0/PictureBox.qml#L123

To make it work with the current implementation
```sh
# remove the current password (will clear the password digest in /etc/shadow, but keep the 'x' in /etc/passwd)
passwd -d username_of_the_paswordless_account

# manually edit /etc/passwd, remove 'x' in password field for the user in question (so that it's actually empty)
vipwd
```
This also requires that you have either `nullok` or `nopasswdlogin` configuration in `/etc/pam.d/sddm`

However, this functionality should be reviewed and reworked so that it 
- either checks the presence of the password digest in the `/etc/shadow` instead (empty password fields in `passwd` are generally not recommended)
- I assume there is a more elegant way of probing if user has a password through `libpam` ?

**Comment by NatTuck** [comment]
@tpecar None of that helps the scenario where passwordless login is desired only via the local display manager while still allowing the normal authentication mechanisms to work otherwise.

**Comment by tpecar** [comment]
@NatTuck Yep, that's true.

I suppose that instead of probing for password presence, sddm should only check whether the user belongs to the `nopasswdlogin` group (or some other group, if this is also used by other services) and set the `model.needsPassword` accordingly.

AFAIK if the group used for determining whether it should be a passwordless login is only used by sddm and within its PAM configuration, other services will still require authentication (which is good & what you're trying to achieve).

But again, the way sddm does detection needs to be reworked to support this.

**Comment by tpecar** [comment]
@NatTuck Also note that the way it does paswordless detection right now seems to have been intentional, see pull request https://github.com/sddm/sddm/pull/1368

**Comment by grahamperrin** [comment]
Is this related? 

KDE bug [440055 – Can't unlock the screen if the user has an empty password](https://bugs.kde.org/show_bug.cgi?id=440055)

**Comment by githubkusi** [comment]
@grahamperrin IMO it's not exactly the same issue. I don't want  an empty passwd for the remote access use case (ssh), I only wan't a password less login when I'm physically at the computer showing sddm. I agree that removing the password at all -could- be a work around (then your linked bug would be related) but it's really not what I was aiming for.

**Comment by NatTuck** [comment]
@tpecar Having a "nopasswdlogin" group that SDDM looks for to allow passwordless logins as you describe sounds great.

**Comment by Crystal4276** [comment]
Passing by one year later...
I'm trying to get rid of the password field in sddm with PAM configured for password less login.

Was it implemented at the end ?

**Comment by alextrical** [comment]
Also interested, as I'm trying to setup a suite of communal laptops for a Fablab

**Comment by likewhoa** [comment]
forgot about this but once I get sometime I will reproduce what I did for Gentoo Linux that allowed password-less while keeping everything outside of it intact.

**Comment by manuvazquez** [comment]
I'm also on Gentoo. Could you please share some hints?

**Comment by auxym** [comment]
I'd also like this feature. My use case is a shared computer for the family in the living room, used to play music, videos and games. I did turn on auto-login, but if anyone logs out, switches to another user or locks the screen, then you need the password to log back in (or reboot, which feels really unnecessary).

I did get passwordless login through PAM by adding the following line add the top of `/etc/pam.d/sddm`:

```
auth        sufficient  pam_succeed_if.so user ingroup nopasswdlogin
```

**Note:** I also had to add the same line to `/etc/pam.d/kde` in order to get passwordless screen unlocking.

And a second note, since I am on openSUSE, I first had to copy both config files from `/usr/lib/pam.d` to `/etc/pam.d`. Hopefully that helps some future googler.

It would be nice to have a single checkbox in Settings, right next to the "auto-login" feature, for this :). Actually I believe that auto-login should also enable passwordless, I don't really see a use case for having auto-login without passwordless. If anyone locks the screen or whatever, you can just reboot the PC to be logged back in.

**Comment by githubkusi** [comment]
> And a second note, since I am on openSUSE, I first had to copy both config files from /usr/lib/pam.d to /etc/pam.d. Hopefully that helps some future googler.

Thanks @auxym , this was the missing step on OpenSuse. For reference /usr/lib/pam.d is not meant to be user-modifiable, /etc/pam.d is (and has precedence over /usr/lib/pam.d). Everything else can be applied as described in the [Arch wiki](https://wiki.archlinux.org/title/SDDM#Passwordless_login)
